Firebase.h

```diff
@@ -25,49 +25,108 @@
 #include <WiFiClientSecure.h>
 #include <ESP8266HTTPClient.h>
 
-// FirebaseError represents a Firebase API error with a code and a
-// message.
-class FirebaseError {
+//TODO(edcoyne) split these into multiple files.
+
+// Result from call to Firebase backend. ALWAYS check isError() before
+// expecting any data.
+class FirebaseResult {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  FirebaseResult(int status);
+  FirebaseResult(int status, const String& response);
+  FirebaseResult(const FirebaseResult& result);
+
+  // True if there was an error completeing call.
+  bool isError() const;
+  String errorMessage() const;
+
+  // True if http status code is 200(OK).
+  bool isOk() const;
+  // Message sent back from Firebase backend.
+  const String& response() const;
+
+  int httpStatus() const {
+    return status_;
   }
+
  private:
-  int _code = 0;
-  String _message = "";
+  int status_;
+  String response_;
 };
 
-// Firebase is the connection to firebase.
+// Low level connection to Firebase backend, you probably want the
+// Firebase class below.
+class FirebaseConnection {
+ public:
+  FirebaseConnection(const String& host);
+  FirebaseConnection& auth(const String& auth);
+
+  const String& host() {
+    return host_;
+  }
+
+  HTTPClient& httpClient(){
+    return http_;
+  }
+
+  String makeURL(const String& path);
+
+  FirebaseResult sendRequest(const char* method, const String& path, const String& value);
+  FirebaseResult sendRequest(const char* method, const String& path);
+
+  FirebaseResult sendRequestGetBody(const char* method, const String& path);
+  FirebaseResult sendRequestGetBody(const char* method, const String& path, const String& value);
+
+ private:
+  HTTPClient http_;
+  const String host_;
+  String auth_;
+};
+
+// Primary client to the Firebase backend.
 class Firebase {
  public:
   Firebase(const String& host);
   Firebase& auth(const String& auth);
-  const FirebaseError& error() const {
-    return _error;
-  }
-  String get(const String& path);
-  String push(const String& path, const String& value);
-  bool connected();
-  Firebase& stream(const String& path);
-  bool available();
+
+  // Fetch result at "path" to a local variable. If the value is too large you will exceed
+  // local memory.
+  FirebaseResult get(const String& path);
+
+  // Add new value to list at "path", will return child name of new item.
+  FirebaseResult push(const String& path, const String& value);
+
+  // Deletes value at "path" from server.
+  FirebaseResult remove(const String& path);
+
+ private:
+  FirebaseConnection connection_;
+};
+
+// Listens on a stream of events from Firebase backend.
+class FirebaseEventStream {
```

Comment:
I'd love if we could keep the stream method attached to the main `Firebase` class (even if the implementation is decoupled), if we decouple the method for making request and getting result it seems that most of the stream interface could apply to regular request, and that the real different is that you can `read()` multiple data from a stream, ie: `availlable()` could also be useful for regular get request to know if there is a body to read.